tests/lib/__init__.py

```diff
@@ -984,95 +985,48 @@ def create_basic_wheel_for_package(
-            License: UNKNOWN
-            Platform: UNKNOWN
-            {requires_dist}
-
-            UNKNOWN
-        """,
-        "{dist_info}/top_level.txt": """
-            {name}
-        """,
-        # Have an empty RECORD because we don't want to be checking hashes.
-        "{dist_info}/RECORD": ""
-    }
+    if extra_files is None:
+        extra_files = {}
 
-    # Some useful shorthands
-    archive_name = "{name}-{version}-py2.py3-none-any.whl".format(
-        name=name, version=version
-    )
-    dist_info = "{name}-{version}.dist-info".format(
-        name=name, version=version
-    )
+    archive_name = "{}-{}-py2.py3-none-any.whl".format(name, version)
+    archive_path = script.scratch_path / archive_name
+
+    requires_dist = depends + [
+        '{package}; extra == "{extra}"'.format(package=package, extra=extra)
+        for extra, packages in extras.items()
+        for package in packages
+    ]
+
+    wheel_builder = make_wheel(
+        name=name,
+        version=version,
+        wheel_metadata_updates={"Tag": ["py2-none-any", "py3-none-any"]},
+        metadata_updates={
+            "Provides-Extra": list(extras),
+            "Requires-Dist": requires_dist,
+        },
+        extra_metadata_files={"top_level.txt": name},
 
-    requires_dist = "\n".join([
-        "Requires-Dist: {}".format(pkg) for pkg in depends
-    ] + [
-        "Provides-Extra: {}".format(pkg) for pkg in extras.keys()
-    ] + [
-        "Requires-Dist: {}; extra == \"{}\"".format(pkg, extra)
-        for extra in extras for pkg in extras[extra]
-    ])
-
-    # Replace key-values with formatted values
-    for key, value in list(files.items()):
-        del files[key]
-        key = key.format(name=name, dist_info=dist_info)
-        files[key] = textwrap.dedent(value).format(
-            name=name, version=version, requires_dist=requires_dist
-        ).strip()
-
-    # Add new files after formatting
-    if extra_files:
-        files.update(extra_files)
-
-    for fname in files:
-        path = script.temp_path / fname
-        path.parent.mkdir(exist_ok=True, parents=True)
-        path.write_bytes(ensure_binary(files[fname]))
-
-    # The base_dir cast is required to make `shutil.make_archive()` use
-    # Unicode paths on Python 2, making it able to properly archive
-    # files with non-ASCII names.
-    retval = script.scratch_path / archive_name
-    generated = shutil.make_archive(
-        retval,
-        'zip',
-        root_dir=script.temp_path,
-        base_dir=text_type(os.curdir),
+        # Have an empty RECORD because we don't want to be checking hashes.
+        record="",
     )
-    shutil.move(generated, retval)
 
-    shutil.rmtree(script.temp_path)
-    script.temp_path.mkdir()
+    # Create the wheel in-memory to add more files.
+    wheel_io = BytesIO(wheel_builder.as_bytes())
+    with ZipFile(wheel_io, "a") as zf:
+        zf.writestr(
+            "{name}/__init__.py".format(name=name),
```

Comment:
Is there a reason you are not using make_wheel's [extra_files](https://github.com/pypa/pip/blob/657cf2515be06e6f935869b5629f11e5045ba987/tests/lib/wheel.py#L308) argument ?

Reply:
I’m… not sure, I likely missed that argument entirely (or confused it with `extra_metadata_files`) and thought there’s not a way to do add arbitrary files. Let me try switching to that and see if the tests pass.

Reply:
That seems to work. Thanks for pointing this out!